File: app/services/checkout_service.rb
require "inttegro"
require "uri"

class CheckoutService
  Checkout = Data.define(:name, :email, :phone, :attempt_id)

  class DemoError < StandardError
    attr_reader :code

    def initialize(code, message)
      @code = code
      super(message)
    end
  end

  class << self
    def parse(values)
      checkout = Checkout.new(
        name: values.fetch("name", "").strip,
        email: values.fetch("email", "").strip,
        phone: values.fetch("phone", "").strip,
        attempt_id: values.fetch("attempt_id", "").strip
      )
      unless checkout.name.present? && checkout.phone.present? && checkout.email.match?(/\A[^\s@]+@[^\s@]+\.[^\s@]+\z/)
        raise DemoError.new("validation_error", "Enter a name, valid email, and phone number.")
      end
      unless checkout.attempt_id.match?(/\A[A-Za-z0-9_-]{8,100}\z/)
        raise DemoError.new("validation_error", "Start a fresh checkout and try again.")
      end
      checkout
    end

    def order_params(checkout, origin)
      {
        request_meta: { idempotency_key: "demo-#{checkout.attempt_id}" },
        customer_data: { name: checkout.name, email_address: checkout.email, phone_number: checkout.phone },
        finalize: true,
        checkout_settings: { redirect_url: "#{origin}/complete", cancel_url: "#{origin}/cancel" },
        line_items: [{
          type: "product",
          product: {
            type: Inttegro::ProductType::DIGITAL,
            name: "Inttegro integration workshop",
            quantity: 1,
            price: Inttegro::PriceParams.new(currency: Inttegro::Money::Currency::GHS, value: 5000)
          }
        }]
      }
    end

    def create(checkout, origin)
      uri = URI.parse(origin)
      unless %w[http https].include?(uri.scheme) && uri.host
        raise DemoError.new("configuration_error", "The demo public URL is invalid.")
      end
      api_key = ENV.fetch("INTTEGRO_API_KEY", "").strip
      raise DemoError.new("configuration_error", "Set INTTEGRO_API_KEY on the server.") if api_key.empty?

      public_origin = "#{uri.scheme}://#{uri.host}"
      public_origin += ":#{uri.port}" unless [80, 443].include?(uri.port)
      order = Inttegro::Client.new(api_key: api_key).orders.create(**order_params(checkout, public_origin))
      checkout_url = order.invoice&.format_value&.web&.url
      raise DemoError.new("api_error", "Inttegro did not return a hosted checkout URL.") if checkout_url.blank?
      { order_id: order.id, checkout_url: checkout_url }
    rescue DemoError
      raise
    rescue Inttegro::APIError
      raise DemoError.new("api_error", "Inttegro rejected the checkout request.")
    rescue StandardError
      raise DemoError.new("api_error", "Checkout is temporarily unavailable.")
    end
  end
end

File: app/Services/CheckoutService.php
<?php

namespace App\Services;

use Inttegro\APIError;
use Inttegro\Client;
use Inttegro\Money\Currency;
use Inttegro\PriceParams;
use Inttegro\ProductType;
use RuntimeException;

/*
 * Inttegro integration map
 *
 * INTTEGRO:FLOW [hosted-checkout] This server-only service creates and
 * finalizes the Kora Market Order and returns its hosted invoice URL.
 * INTTEGRO:SECURITY [server-api-key] INTTEGRO_API_KEY belongs only in the
 * trusted PHP process. Never expose it through VITE_ variables, Blade, public
 * assets, logs, or error responses.
 * INTTEGRO:ALTERNATIVE [hosted-checkout] Direct API payment offers more UI
 * control but makes the merchant own more payment state, recovery,
 * method-specific behavior, testing, and compliance analysis.
 * INTTEGRO:DOCS https://studio.inttegro.com/accept-payment-with-inttegro-checkout
 * INTTEGRO:DOCS https://studio.inttegro.com/orders
 * INTTEGRO:DOCS https://studio.inttegro.com/keys
 * See ../../../INTEGRATION_GUIDE.md and ../../../integration-decisions.json for the shared
 * rationale and machine-readable alternatives.
 */

final class DemoError extends RuntimeException
{
    public function __construct(public readonly string $errorCode, string $message, ?\Throwable $previous = null)
    {
        parent::__construct($message, 0, $previous);
    }
}

final class CheckoutService
{
    /** @param array{name: string, email: string, phone: string, attempt_id: string} $checkout */
    public static function orderPayload(array $checkout, string $origin): array
    {
        return [
            // INTTEGRO:DECISION [stable-idempotency-key] Reuse the key for a
            // retry of this logical checkout. Production code should persist a
            // cart/invoice-derived key; creating a new key after a timeout can
            // create a duplicate. https://studio.inttegro.com/idempotency
            'request_meta' => ['idempotency_key' => 'demo-'.$checkout['attempt_id']],
            // INTTEGRO:DECISION [inline-customer] customer_data suits this guest
            // flow. Account-based apps should resolve customer_id server-side;
            // the Orders API accepts exactly one customer representation.
            'customer_data' => ['name' => $checkout['name'], 'email_address' => $checkout['email'], 'phone_number' => $checkout['phone']],
            // INTTEGRO:DECISION [finalize-on-create] The cart is settled, so one
            // operation freezes the Order and produces invoice formats. Use a
            // draft, updates, then finalization when tax, shipping, inventory,
            // approval, or line items can still change.
            'finalize' => true,
            'checkout_settings' => ['redirect_url' => $origin.'/complete', 'cancel_url' => $origin.'/cancel'],
            'line_items' => [[
                'type' => 'product',
                'product' => [
                    // INTTEGRO:DECISION [inline-product] Inline product data
                    // keeps this demo self-contained. Catalog merchants can use
                    // product_id plus price or price_id; do not mix both shapes.
                    'type' => ProductType::Physical,
                    'name' => 'Dawn Brew Set',
                    'quantity' => 1,
                    // INTTEGRO:DECISION [minor-unit-money] Integer 5000 is
                    // GHS 50.00. Use a currency-aware decimal/money type when
                    // converting variable amounts.
                    'price' => new PriceParams(Currency::GHS, 5000),
                ],
            ]],
        ];
    }

    /** @param array{name: string, email: string, phone: string, attempt_id: string} $checkout
     *  @return array{order_id: string, checkout_url: string}
     */
    public static function create(array $checkout, string $origin): array
    {
        // INTTEGRO:SECURITY [server-api-key] Load the key from a deployment
        // secret store in production and rotate it there. Do not serialize it.
        $apiKey = trim((string) getenv('INTTEGRO_API_KEY'));
        if ($apiKey === '') {
            throw new DemoError('configuration_error', 'Set INTTEGRO_API_KEY on the server.');
        }
        // INTTEGRO:SECURITY [configured-public-origin] Prefer an allow-listed
        // origin in production. Request-derived values are safe only behind a
        // precise trusted-proxy configuration. Return URLs must be HTTP(S), and
        // Inttegro appends order_id after checkout.
        $parts = parse_url($origin);
        if (!is_array($parts) || !in_array($parts['scheme'] ?? '', ['http', 'https'], true) || empty($parts['host'])) {
            throw new DemoError('configuration_error', 'The demo public URL is invalid.');
        }
        $publicOrigin = $parts['scheme'].'://'.$parts['host'].(isset($parts['port']) ? ':'.$parts['port'] : '');

        try {
            // INTTEGRO:ALTERNATIVE [server-api-key] A production container may
            // inject one startup-configured Client for transport reuse and
            // application-owned OpenTelemetry. Per-call construction is concise
            // for the demo. https://studio.inttegro.com/sdk-observability
            $order = (new Client($apiKey))->orders->create(self::orderPayload($checkout, $publicOrigin));
            // INTTEGRO:DECISION [returned-checkout-url] Use the URL in the
            // response. Constructing one from the order ID relies on
            // undocumented routing conventions.
            $checkoutUrl = $order->invoice?->format?->web?->url;
            if (!$checkoutUrl) {
                throw new DemoError('api_error', 'Inttegro did not return a hosted checkout URL.');
            }
            return ['order_id' => $order->id, 'checkout_url' => $checkoutUrl];
        } catch (DemoError $error) {
            throw $error;
        } catch (APIError $error) {
            // INTTEGRO:SECURITY [safe-error-boundary] Keep raw API bodies,
            // credentials, and traces in protected diagnostics. Public messages
            // remain stable; log only bounded metadata and request IDs.
            throw new DemoError('api_error', 'Inttegro rejected the checkout request.', previous: $error);
        } catch (\Throwable $error) {
            throw new DemoError('api_error', 'Checkout is temporarily unavailable.', previous: $error);
        }
    }
}
